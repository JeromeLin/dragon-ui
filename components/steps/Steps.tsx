import React, { PureComponent, Children, cloneElement } from 'react';
import classnames from 'classnames';
import StepsItem from './StepsItem';
import { StepsProps } from './PropsType';

export default class Steps extends PureComponent<StepsProps> {
  static Item: typeof StepsItem;

  static defaultProps = {
    prefixCls: 'zw-steps',
    direction: 'horizontal',
    current: 0,
    status: 'process',
  };

  render() {
    const {
      prefixCls,
      className,
      style,
      direction,
      children,
      current,
      onChange,
      status,
    } = this.props;

    const classStr = classnames(
      prefixCls,
      className,
      `${prefixCls}--${direction}`,
    );

    const filteredChildren = Children.toArray(children).filter((c) => !!c);
    const lastIndex = filteredChildren.length - 1;

    return (
      <div className={classStr} style={style}>
        {Children.map(filteredChildren, (child: React.ReactElement, index: number) => {
          if (!child) {
            return null;
          }
          const childProps = {
            prefixCls,
            disabled: false,
            stepIndex: index,
            current,
            isLastStep: lastIndex === index,
            status: current === index ? status : 'wait',
            onClick(currentIndex: number) {
              onChange && onChange(currentIndex);
            },
            stepClick: typeof onChange === 'function',
            ...child.props,
          };
          return cloneElement(child, childProps);
        })}
      </div>
    );
  }
}
