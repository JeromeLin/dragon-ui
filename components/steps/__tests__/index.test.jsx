
import React from 'react';
import { render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Steps from '../index';

describe('Steps', () => {
  it('renders normal steps correctly', () => {
    const wrapper = render(
      <div>
        <Steps current={2}>
          <Steps.Item title="Finished" />
          <Steps.Item title="In Progress" />
          <Steps.Item title="Waiting" />
        </Steps>
      </div>,
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('click steps', () => {
    const onChange = jest.fn();
    const wrapper = mount(
      <div>
        <Steps onChange={onChange}>
          <Steps.Item title="Finished" />
          <Steps.Item title="In Progress" disabled />
          <Steps.Item title="Waiting" />
        </Steps>
      </div>,
    );
    wrapper.find('.zw-steps-item').at(0).simulate('click');
    wrapper.find('.zw-steps-item').at(1).simulate('click');
    wrapper.find('.zw-steps-item').at(2).simulate('click');

    expect(onChange).toBeCalledTimes(2);
    expect(onChange).toBeCalledWith(2);
  });

  it('steps: direction and description', () => {
    const wrapper = mount(
      <div>
        <Steps current={1} direction="vertical" status="error">
          <Steps.Item title="Finished" description="description" />
          <Steps.Item title="In Progress" />
          <Steps.Item title="Waiting" />
        </Steps>
      </div>,
    );

    expect(wrapper.find('.zw-steps').hasClass('zw-steps--vertical')).toBeTruthy();
    expect(wrapper.find('.zw-steps-item').at(1).hasClass('zw-steps-item--error')).toBeTruthy();
  });
});
