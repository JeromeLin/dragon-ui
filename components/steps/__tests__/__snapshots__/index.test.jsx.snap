// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Steps renders normal steps correctly 1`] = `
<div>
  <div
    class="zw-steps zw-steps--horizontal"
  >
    <div
      class="zw-steps-item zw-steps-item--finish"
    >
      <div
        class="zw-steps-item__icon"
      >
        <i
          class="zw-icon zw-icon--right zw-icon--success zw-icon--sm"
          theme="success"
          type="right"
        >
          <svg
            fill="currentColor"
            height="1em"
            viewBox="0 0 32 32"
            width="1em"
          >
            <use
              href="#right"
            />
          </svg>
        </i>
      </div>
      <div
        class="zw-steps-item__tail"
      />
      <div
        class="zw-steps-item__content"
      >
        <div
          class="zw-steps-item__title"
        >
          Finished
        </div>
      </div>
    </div>
    <div
      class="zw-steps-item zw-steps-item--finish"
    >
      <div
        class="zw-steps-item__icon"
      >
        <i
          class="zw-icon zw-icon--right zw-icon--success zw-icon--sm"
          theme="success"
          type="right"
        >
          <svg
            fill="currentColor"
            height="1em"
            viewBox="0 0 32 32"
            width="1em"
          >
            <use
              href="#right"
            />
          </svg>
        </i>
      </div>
      <div
        class="zw-steps-item__tail"
      />
      <div
        class="zw-steps-item__content"
      >
        <div
          class="zw-steps-item__title"
        >
          In Progress
        </div>
      </div>
    </div>
    <div
      class="zw-steps-item zw-steps-item--active zw-steps-item--process"
    >
      <div
        class="zw-steps-item__icon"
      >
        3
      </div>
      <div
        class="zw-steps-item__content"
      >
        <div
          class="zw-steps-item__title"
        >
          Waiting
        </div>
      </div>
    </div>
  </div>
</div>
`;
